maps/time_key_map.go

```diff
@@ -0,0 +1,133 @@
+package maps
+
+import (
+	"time"
+
+	"github.com/akitasoftware/go-utils/optionals"
+	"github.com/akitasoftware/go-utils/sets"
+	"github.com/akitasoftware/go-utils/slices"
+)
+
+func getInternalMapKey(t time.Time) int64 {
+	return t.Unix()
+}
+
+func getReverseKey(key int64) time.Time {
+	return time.Unix(key, 0)
+}
+
+type TimeMap[V any] struct {
+	internalMap Map[int64, V]
+}
+
+func NewTimeMap[V any]() TimeMap[V] {
+	return TimeMap[V]{
+		internalMap: NewMap[int64, V](),
+	}
+}
+
+func (m TimeMap[V]) Put(k time.Time, v V) {
+	key := getInternalMapKey(k)
+	m.internalMap.Put(key, v)
+}
+
+func (m TimeMap[V]) Upsert(k time.Time, v V, onConflict func(v, newV V) V) {
+	key := getInternalMapKey(k)
+	m.internalMap.Upsert(key, v, onConflict)
+}
+
+// If the key k is not already in the map, then it is entered into the map with
+// the value v.
+func (m TimeMap[V]) PutIfAbsent(k time.Time, v V) {
+	m.GetOrValue(k, v)
```

Comment:
**Blocking.** This is a copy of the implementation in `Map`. We should instead convert the key and call `Map.PutIfAbsent`. Similarly for `ComputeIfAbsent` and `ComputeIfAbsentNoError`.
```suggestion
	key := getInternalMapKey(k)
	m.internalMap.PutIfAbsent(key, v)
```

Reply:
Shouldn't we try to stay close to the Map's implementation? This would help us maintain consistency across the different Maps we add. 
I prefer using `GetOrValue` to add custom logic and then delegating to `Map.` Thoughts?

Reply:
For situations like this, it's generally better to delegate to an underlying implementation than to copy it, for a couple of reasons.
* Most other methods are following the same pattern. Internal consistency is generally preferred, unless there is strong reason to deviate. (For example, `TimeMap.KeySet` has its own implementation, to avoid iterating through the key set twice.)
* If we end up changing the implementation of `Map.PutIfAbsent`, then we only need to change that one place, and `TimeMap` would benefit.